metricq_wizard_backend/metricq/cluster_scanner.py

```diff
@@ -0,0 +1,521 @@
+        assert self.lock is not None
+        return self.lock.locked()
+
+    async def run_once(self) -> None:
+        assert self.lock is not None
+        if self.lock.locked():
+            raise RuntimeError("Scan already running")
+
+        async with self.lock:
+            logger.warn("Starting Cluster Health Scan")
+            try:
+                await self._run_scan()
+            except Exception as e:
+                logger.error(f"Cluster Scan failed: {e}")
+                logger.error("".join(traceback.format_exception(e)))
+            finally:
+                logger.warn("Cluster Health Scan Finished")
+
+    async def _run_scan(self) -> None:
+        # TODO add check for (db) queue bindings
+        # TODO add check for token names
+
+        assert self.db_metadata is not None
+
+        async with HistoryClient(self.token, self.url, add_uuid=True) as client:
+            tasks = AsyncTaskPool(max_tasks=250)
+
+            async for doc in self.db_metadata.docs():
+                metric = doc.id
+                metadata = doc.data
+
+                # this assert is purely for mypy. doc.data can only return None
+                # is the doc does not exist. Which it clearly does, as we only
+                # get existing documents from the docs iterator.
+                assert metadata is not None
+
+                await tasks.append(self.check_metric_metadata(metric, metadata))
+
+                if metadata.get("historic", False):
+                    # Only check the db status for historic metrics
+                    await tasks.append(
+                        self.check_metric_is_dead(client, metric, metadata)
+                    )
+                    await tasks.append(
+                        self.check_metric_for_infinites(client, metric, metadata)
+                    )
+
+                # there is no tooling for renaming metrics, so bad
+                # names is nothing we should warn about yet.
+                # await tasks.append(self.check_metric_name(metric))
+
+            await tasks.completed()
+
+    async def create_issue_report(
+        self,
+        type: str,
+        scope_type: str,
+        scope: str,
+        date: str | None = None,
+        severity: str | None = None,
+        **kwargs: Any,
+    ):
+        assert self.db_issues is not None
+        report = await self.db_issues.create(
+            id=f"{type}-{scope_type}-{scope}", exists_ok=True
+        )
+
+        report["severity"] = "warning" if severity is None else severity
+
+        if "first_detection_date" not in report:
+            # only set first_detection_date when creating the report, not
+            # on updates
+            report["first_detection_date"] = (
+                str(Timestamp.now().datetime.isoformat()
+                    ) if date is None else date
+            )
+
+        report["date"] = (
+            str(Timestamp.now().datetime.isoformat()) if date is None else date
+        )
```

Comment:
It's pretty dangerous to take date as string here, we won't have any guarantee that callers use isoformat and not "Yesterday".